app/upgrades_test.go

```diff
@@ -0,0 +1,72 @@
+package app
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+
+	v4 "github.com/realiotech/realio-network/app/upgrades/v4"
+)
+
+func TestV4Upgrade(t *testing.T) {
+	app := Setup(false, nil, 4)
+	ctx := app.BaseApp.NewContext(false, tmproto.Header{Height: app.LastBlockHeight() + 1})
+	validators := app.StakingKeeper.GetAllValidators(ctx)
+
+	comm0 := stakingtypes.CommissionRates{
+		Rate:          sdk.MustNewDecFromStr("0.01"),
+		MaxRate:       sdk.MustNewDecFromStr("0.01"),
+		MaxChangeRate: sdk.MustNewDecFromStr("0.02"),
+	}
+	comm1 := stakingtypes.CommissionRates{
+		Rate:          sdk.MustNewDecFromStr("0.02"),
+		MaxRate:       sdk.MustNewDecFromStr("0.03"),
+		MaxChangeRate: sdk.MustNewDecFromStr("0.02"),
+	}
+	comm2 := stakingtypes.CommissionRates{
+		Rate:          sdk.MustNewDecFromStr("0.06"),
+		MaxRate:       sdk.MustNewDecFromStr("0.07"),
+		MaxChangeRate: sdk.MustNewDecFromStr("0.02"),
+	}
+	comm3 := stakingtypes.CommissionRates{
+		Rate:          sdk.MustNewDecFromStr("0.1"),
+		MaxRate:       sdk.MustNewDecFromStr("0.2"),
+		MaxChangeRate: sdk.MustNewDecFromStr("0.1"),
+	}
+
+	validators[0].Commission.CommissionRates = comm0
+	validators[1].Commission.CommissionRates = comm1
+	validators[2].Commission.CommissionRates = comm2
+	validators[3].Commission.CommissionRates = comm3
+
+	app.StakingKeeper.SetValidator(ctx, validators[0])
+	app.StakingKeeper.SetValidator(ctx, validators[1])
+	app.StakingKeeper.SetValidator(ctx, validators[2])
+	app.StakingKeeper.SetValidator(ctx, validators[3])
+
+	upgradePlan := upgradetypes.Plan{
+		Name:   v4.UpgradeName,
+		Height: ctx.BlockHeight(),
+	}
+	err := app.UpgradeKeeper.ScheduleUpgrade(ctx, upgradePlan)
```

Comment:
can you have some vals with above minComm commission rate, also set staking min comm rate to be the `NewMinCommisionRate`